src/System.Private.CoreLib/src/Internal/Runtime/Augments/EnvironmentAugments.Unix.cs

```diff
@@ -30,7 +30,89 @@ private static void SetEnvironmentVariableCore(string variable, string value)
             if ("".Length != 0)
                 throw new NotImplementedException(); // Need to return something better than an empty environment block.
 
-            return Array.Empty<KeyValuePair<string,string>>();
+            unsafe
+            {
+                // Get byte** of environment variables from native interop
+                var unsafeBlock = (byte**)Interop.Sys.EnvironGetSystemEnvironment();
+                if (unsafeBlock == (byte**)0)
+                    throw new OutOfMemoryException();
+                
+                // Find total length of two-dimensional byte**
+                int rowIndex = 0;
+                int totalLength = 0;
+                while (unsafeBlock[rowIndex] != null && unsafeBlock[rowIndex][0] != 0)
+                {
+                    byte* p = unsafeBlock[rowIndex];
+                    while (*p != 0) // Continue until end-of-line char '\0'
+                    {
+                        p++;
+                    }
+                    totalLength += checked((int)(p - unsafeBlock[rowIndex] + 1));
+                    rowIndex++;
+                }
+
+                // Copy two-dimensional byte** to a flat char[] for parsing
+                rowIndex = 0;
+                var blockIndex = 0;
+                char[] block = new char[totalLength];
+                while (unsafeBlock[rowIndex] != null && unsafeBlock[rowIndex][0] != 0)
+                {
+                    byte* p = unsafeBlock[rowIndex];
+                    while (*p != 0) // Continue until end-of-line char '\0'
+                    {
+                        p++;
+                    }
+                    var rowLength = checked((int)(p - unsafeBlock[rowIndex] + 1));
+                    for (int i = 0; i < rowLength; i++)
+                    {
+                        // Copy original byte to a Unicode char in our flat char[]
+                        block[blockIndex++] = Convert.ToChar(unsafeBlock[rowIndex][i]);
```

Comment:
We use UTF8 encoding to convert OS strings to .NET strings on Unix.

Reply:
Together with all the suggestions, this can be something like (have not tested it):

```csharp
var block = (byte**)Interop.Sys.GetEnviron();
if (block == null)
    return;

for (byte** cur = block; *cur != null; cur++)
{
    byte* entry = *cur;
    byte* p = entry;

    // Skip the key. On some old OS, the environment block can be corrupted.
    // Some will not have '=', so we need to check for '\0'. 
    while (*p != '=' && *p != '\0')
        p++;

    // Skip over environment variables starting with '=' and environment variables with no value 
    if (p == entry || *p == '\0')
        continue;

    string key = new string((sbyte*)entry, 0, checked(p - entry));
    string value = new string((sbyte*)(p + 1));

    yield return new KeyValuePair<string, string>(key, value);
}
```

Reply:
Done via rework.